clickhouse-admin/src/context.rs

```diff
@@ -3,26 +3,50 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use crate::{ClickhouseCli, Clickward};
+
+use anyhow::{anyhow, bail, Result};
+use camino::Utf8PathBuf;
+use clickhouse_admin_types::{
+    CLICKHOUSE_KEEPER_CONFIG_DIR, CLICKHOUSE_KEEPER_CONFIG_FILE,
+    CLICKHOUSE_SERVER_CONFIG_DIR, CLICKHOUSE_SERVER_CONFIG_FILE,
+};
 use omicron_common::address::CLICKHOUSE_TCP_PORT;
+use omicron_common::api::external::Generation;
 use oximeter_db::Client as OximeterClient;
 use slog::Logger;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
 use std::net::SocketAddrV6;
+use std::str::FromStr;
 use std::sync::Arc;
 use tokio::sync::Mutex;
 
 pub struct KeeperServerContext {
     clickward: Clickward,
     clickhouse_cli: ClickhouseCli,
     log: Logger,
+    pub generation: Mutex<Option<Generation>>,
 }
 
 impl KeeperServerContext {
-    pub fn new(clickhouse_cli: ClickhouseCli) -> Self {
-        let log = clickhouse_cli
-            .log
-            .new(slog::o!("component" => "KeeperServerContext"));
+    pub fn new(
+        log: &Logger,
+        binary_path: Utf8PathBuf,
+        listen_address: SocketAddrV6,
+    ) -> Result<Self> {
+        let clickhouse_cli =
+            ClickhouseCli::new(binary_path, listen_address, log);
+        let log = log.new(slog::o!("component" => "KeeperServerContext"));
         let clickward = Clickward::new();
-        Self { clickward, clickhouse_cli, log }
+        let config_path = Utf8PathBuf::from_str(CLICKHOUSE_KEEPER_CONFIG_DIR)
+            .unwrap()
+            .join(CLICKHOUSE_KEEPER_CONFIG_FILE);
+
+        // If there is already a configuration file with a generation number we'll
+        // use that. Otherwise, we set the generation number to None.
+        let gen = read_generation_from_file(config_path)?;
+        let generation = Mutex::new(gen);
```

Comment:
It's become practice at Oxide to avoid tokio mutexes wherever possible as they have significant problems when cancelled and generally just don't do what we want. I realize there's already some usage here with regards to initialization. We don't have to fix that in this PR, but we should avoid adding new uses. We should instead use a `std::sync::mutex`. I left a comment below about this as well.

See the following for more details:
https://rfd.shared.oxide.computer/rfd/0400#no_mutex
https://rfd.shared.oxide.computer/rfd/0397#_example_with_mutexes

Reply:
lol I was definitely on the fence on that one, I went for consistency in the end https://github.com/oxidecomputer/omicron/pull/7347/commits/be1afc76d9b0a5393c95a389f4af22784b67217d#diff-c816600501b7aaa7de4a2eb9dc86498662030cea6390fa23e11a22c990efb510L28-L29

Thanks for the links! Hadn't seen those RFDs, will read them both